src/Auxiliary/Aux_Record_Center.cpp

```diff
@@ -0,0 +1,163 @@
+// Return      :  None
+//-------------------------------------------------------------------------------------------------------
+void Aux_Record_Center()
+{
+
+   const char FileName[] = "Record__Center";
+   static bool FirstTime = true;
+
+// 1. Maximum fluid density in HYDRO/ELBDM
+   Extrema_t Max_Dens;
+   Max_Dens.Field     = _DENS;
+   Max_Dens.Radius    = HUGE_NUMBER; // entire domain
+   Max_Dens.Center[0] = amr->BoxCenter[0];
+   Max_Dens.Center[1] = amr->BoxCenter[1];
+   Max_Dens.Center[2] = amr->BoxCenter[2];
+
+   Aux_FindExtrema( &Max_Dens, EXTREMA_MAX, 0, TOP_LEVEL, PATCH_LEAF );
+
+
+#  ifdef PARTICLE
+// 2. Maximum particle density
+   Extrema_t Max_ParDens;
+   Max_ParDens.Field     = _PAR_DENS;
+   Max_ParDens.Radius    = HUGE_NUMBER; // entire domain
+   Max_ParDens.Center[0] = amr->BoxCenter[0];
+   Max_ParDens.Center[1] = amr->BoxCenter[1];
+   Max_ParDens.Center[2] = amr->BoxCenter[2];
+
+   Aux_FindExtrema( &Max_ParDens, EXTREMA_MAX, 0, TOP_LEVEL, PATCH_LEAF );
+
+
+// 3. Maximun total density including fluid density and particle density
+   Extrema_t Max_TotDens;
+   Max_TotDens.Field     = _TOTAL_DENS;
+   Max_TotDens.Radius    = HUGE_NUMBER; // entire domain
+   Max_TotDens.Center[0] = amr->BoxCenter[0];
+   Max_TotDens.Center[1] = amr->BoxCenter[1];
+   Max_TotDens.Center[2] = amr->BoxCenter[2];
+
+   Aux_FindExtrema( &Max_TotDens, EXTREMA_MAX, 0, TOP_LEVEL, PATCH_LEAF );
+#  endif
+
+
+#  ifdef GRAVITY
+// 4. Minimun gravitational potential
+   Extrema_t Min_Pote;
+   Min_Pote.Field     = _POTE;
+   Min_Pote.Radius    = HUGE_NUMBER; // entire domain
+   Min_Pote.Center[0] = amr->BoxCenter[0];
+   Min_Pote.Center[1] = amr->BoxCenter[1];
+   Min_Pote.Center[2] = amr->BoxCenter[2];
+
+   Aux_FindExtrema( &Min_Pote, EXTREMA_MIN, 0, TOP_LEVEL, PATCH_LEAF );
+#  endif
+
+
+// 5. Center of mass for the total density field
+// set an initial guess by the peak density position or the user-specified center
+   double CoM_ref[3];
+   if ( COM_CEN_X < 0.0  ||  COM_CEN_Y < 0.0  ||  COM_CEN_Z < 0.0 )
+   {
+#     ifdef PARTICLE
+      for (int d=0; d<3; d++) CoM_ref[d] = Max_TotDens.Coord[d];
+#     else
+      for (int d=0; d<3; d++) CoM_ref[d] = Max_Dens.Coord[d];
+#     endif
+   }
+   else
+   {
+      CoM_ref[0] = COM_CEN_X;
+      CoM_ref[1] = COM_CEN_Y;
+      CoM_ref[2] = COM_CEN_Z;
+   }
+
+// find the center of mass
+   double CoM_Coord[3];
+   double FinaldR;
+   int    FinalNIter;
+   Aux_FindWeightedAverageCenter( CoM_Coord, CoM_ref, COM_MAX_R, COM_MIN_RHO, _TOTAL_DENS, COM_TOL_ERR_R, COM_N_ITER_MAX, &FinaldR, &FinalNIter );
+
+
+// Output the center to file
+   if ( MPI_Rank == 0 )
+   {
+//    Output the header
+      if ( FirstTime )
+      {
+         if ( Aux_CheckFileExist(FileName) )
+            Aux_Message( stderr, "WARNING : file \"%s\" already exists !!\n", FileName );
+         else
+         {
+            FILE *File = fopen( FileName, "w" );
+            fprintf( File, "#%19s  %10s", "Time", "Step" );
```

Comment:
Do you consider combining the separated `fprintf()` here, and the two `fprintf()` within the `PARTICLE` block?